clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1581,86 +1581,131 @@ class SyclKernelFieldChecker : public SyclKernelFieldHandler {
+
+    return false;
   }
 
-  void checkAccessorType(QualType Ty, SourceRange Loc) {
+  bool checkAccessorType(QualType Ty, SourceRange Loc) {
     assert(Util::isSyclAccessorType(Ty) &&
            "Should only be called on SYCL accessor types.");
 
     const RecordDecl *RecD = Ty->getAsRecordDecl();
+    bool isInvalid = false;
     if (const ClassTemplateSpecializationDecl *CTSD =
             dyn_cast<ClassTemplateSpecializationDecl>(RecD)) {
       const TemplateArgumentList &TAL = CTSD->getTemplateArgs();
+      if (TAL.size() < 5)
+        return SemaRef.Diag(
+            Loc.getBegin(),
+            diag::err_sycl_invalid_accessor_template_param_number);
+
+      // A typename specifying the data type that the accessor
+      // is providing access to. Ensure SYCL type restrictions
+      // are applied.
       TemplateArgument TA = TAL.get(0);
-      const QualType TemplateArgTy = TA.getAsType();
+      llvm::DenseSet<QualType> Visited;
+      checkSYCLType(SemaRef, TA.getAsType(), Loc, Visited);
+
+      // An integer specifying the dimensionality of the accessor.
+      TA = TAL.get(1);
+      if (TA.getKind() != TemplateArgument::Integral)
+        isInvalid = SemaRef.Diag(Loc.getBegin(),
+                                 diag::err_sycl_invalid_accessor_template_param)
+                    << /*third*/ 0 << /*access::mode*/ 0;
+
+      // A value of access::mode specifying the mode of access
+      // the accessor is providing.
+      TA = TAL.get(2);
+      if (TA.getKind() != TemplateArgument::Integral ||
+          !TA.getIntegralType()->getAs<EnumType>())
+        isInvalid = SemaRef.Diag(Loc.getBegin(),
+                                 diag::err_sycl_invalid_accessor_template_param)
+                    << /*third*/ 1 << /*access::mode*/ 1;
+
+      // A value of access::target specifying the target of access
+      // the accessor is providing
+      TA = TAL.get(3);
+      if (TA.getKind() != TemplateArgument::Integral ||
+          !TA.getIntegralType()->getAs<EnumType>())
+        isInvalid = SemaRef.Diag(Loc.getBegin(),
+                                 diag::err_sycl_invalid_accessor_template_param)
+                    << /*fourth*/ 2 << /*access::target*/ 2;
+
+      // A value of access::placeholder specifying whether the
+      // accessor is a placeholder accessor.
+      TA = TAL.get(4);
+      if (TA.getKind() != TemplateArgument::Integral ||
+          !TA.getIntegralType()->getAs<EnumType>())
+        isInvalid = SemaRef.Diag(Loc.getBegin(),
+                                 diag::err_sycl_invalid_accessor_template_param)
+                    << /*fifth*/ 3 << /*access::placeholder*/ 3;
 
       if (TAL.size() > 5)
-        checkPropertyListType(TAL.get(5), Loc.getBegin());
-      llvm::DenseSet<QualType> Visited;
-      checkSYCLType(SemaRef, TemplateArgTy, Loc, Visited);
+        if (checkPropertyListType(TAL.get(5), Loc.getBegin()))
+          isInvalid = true;
```

Comment:
```suggestion
        isInvalid = checkPropertyListType(TAL.get(5), Loc.getBegin());
```

Reply:
What happens if there's even more arguments? We should probably diagnose them as being unexpected? Or should this be a `while` loop for all the remaining arguments?